etcd/src/main/scala/io/buoyant/etcd/Key.scala

```diff
@@ -0,0 +1,210 @@
+   * already exist.
+   */
+  def set(
+    value: Option[Buf],
+    ttl: Option[Duration] = None,
+    prevExist: Boolean = false
+  ): Future[NodeOp] = {
+    var params = noParams
+    value match {
+      case Some(Buf.Utf8(value)) =>
+        params = params :+ "value" -> value
+
+      case _ =>
+        params = params :+ "dir" -> "true"
+    }
+    for (ttl <- ttl) {
+      val v = if (ttl == Duration.Zero) "" else ttl.inSeconds.toString
+      params = params :+ "ttl" -> v
+      if (prevExist) {
+        params = params :+ "prevExist" -> "true"
+      }
+    }
+
+    val req = mkReq(keysPrefixPath ++ key, Method.Put, params)
+    client(req) flatMap (toNodeOp(req, _, key, params))
+  }
+
+  def create(
+    value: Option[Buf],
+    ttl: Option[Duration] = None
+  ): Future[NodeOp] = {
+    var params = noParams
+    value match {
+      case Some(Buf.Utf8(value)) =>
+        params = params :+ ("value" -> value)
+      case None =>
+        params = params :+ ("dir" -> "true")
+    }
+    for (ttl <- ttl) {
+      params = params :+ ("ttl" -> ttl.inSeconds.toString)
+    }
+
+    val req = mkReq(keysPrefixPath ++ key, Method.Post, params)
+    client(req) flatMap (toNodeOp(req, _, key, params))
+  }
+
+  def compareAndSwap(
+    value: Buf,
+    prevIndex: Option[Long] = None,
+    prevValue: Option[Buf] = None,
+    prevExist: Boolean = true
+  ): Future[NodeOp] = {
+    require(prevIndex.isDefined || prevValue.isDefined || !prevExist)
+
+    var params = noParams
+    for (i <- prevIndex) {
+      params = params :+ "prevIndex" -> i.toString
+    }
+    for (Buf.Utf8(v) <- prevValue) {
+      params = params :+ "prevValue" -> v
+    }
+    if (!prevExist) {
+      params = params :+ "prevExist" -> "false"
+    }
+    val Buf.Utf8(v) = value
+    params = params :+ "value" -> v
+
+    val req = mkReq(keysPrefixPath ++ key, Method.Put, params)
+    client(req) flatMap (toNodeOp(req, _, key, params))
+  }
+
+  def delete(
+    dir: Boolean = false,
+    recursive: Boolean = false
+  ): Future[NodeOp] = {
+    var params = noParams
+    if (dir) {
+      params = params :+ ("dir" -> "true")
+      if (recursive) {
+        params = params :+ ("recursive" -> "true")
+      }
+    }
+
+    val req = mkReq(keysPrefixPath ++ key, Method.Delete, params)
+    client(req) flatMap (toNodeOp(req, _, key, params))
+  }
+
+  private[this] def getIndex(node: Node): Long = node match {
+    case Data(_, idx, _, _, _) => idx
+    case Dir(_, idx, _, _, nodes) => nodes.foldLeft(idx)(_ max getIndex(_))
+  }
+
+  def events(
+    recursive: Boolean = false,
+    backoff: Stream[Duration] = Stream.empty
+  ): Event[Try[NodeOp]] = new Event[Try[NodeOp]] {
+    private[this] val origBackoff = backoff
+
+    def register(witness: Witness[Try[NodeOp]]) = {
+      @volatile var closing = false
+      @volatile var currentOp: Future[NodeOp] = Future.never
+
+      def loop(idx: Option[Long], backoff: Stream[Duration]): Unit =
```

Comment:
I've documented this, but I think this is correct given the semantics of the API